### PR TITLE
Feature/update widgets and fields

### PR DESCRIPTION
**Pull request by AltusBarry**
Add widgets for each of the reCAPTCHA types we currently support.
Add templates for each widget.
Each widget now passes a UUID to its templates.
Updated README
Add tests for some sanity checking.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -39,40 +34,39 @@ Installation
...
-   These can also be specificied per field by passing the ``public_key`` or
-   ``private_key`` parameters to ``ReCaptchaField`` - see field usage below.
+        INSTALLED_APPS = [
+            ...,
+            'captcha',
+            ...
+        ]
 
-#. By default the reCAPTCHA V2 -Checkbox widget will be used, to make use of the reCAPTCHA V2 - Invisible widget add the following setting:
+#. Add the keys reCAPTCHA have given you to your Django production settings (leave development settings blank to use the default test keys) as
```
Comment (cobusc) [inline README.rst]: "has given"

Comment (cobusc) [reply README.rst]: How about:
`Add the keys reCAPTCHA has given you to your Django production settings. Note that omitting these settings will default to a set of test keys which can be used for development.`
The sentence above seems too long with the part in brackets.

File: README.rst [hunk 2]
```diff
@@ -39,40 +34,39 @@ Installation
...
 -----
 
-Field
-~~~~~
+Fields
+~~~~~~
 
 The quickest way to add reCAPTCHA to a form is to use the included
-``ReCaptchaField`` field class. A ``ReCaptcha`` widget will be rendered with
+``ReCaptchaField`` field class. A ``ReCaptchaV2Checkbox`` widget will be rendered with
```
Comment (cobusc) [inline README.rst]: "...will be rendered by default..."

File: README.rst [hunk 3]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
         private_key='98dfg6df7g56df6gdfgdfg65JHJH656565GFGFGs',
     )
 
-If specified these parameters will be used instead of your reCAPTCHA project
-settings.
+If specified these parameters will be used instead of your reCAPTCHA project settings.
```
Comment (cobusc) [inline README.rst]: "If specified, ..."

File: README.rst [hunk 4]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
...
+
+Widgets
+~~~~~~~
+
+There are two widgets that can be used with the ``ReCaptchaField``;
```
Comment (cobusc) [inline README.rst]: ":" instead of ";"

File: README.rst [hunk 5]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
...
+
+    ``ReCaptchaV2Checkbox`` for Google reCAPTCHA V2 - Checkbox
```
Comment (cobusc) [inline README.rst]: It would be nice to make "Google reCAPTCHA V2 - Checkbox" a link to Google's documentation for this.

File: README.rst [hunk 6]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
...
+
+    ``ReCaptchaV2Invisible`` for Google reCAPTCHA V2 - Invisible
```
Comment (cobusc) [inline README.rst]: A link would be nice here too.

File: README.rst [hunk 7]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
...
+    captcha = fields.ReCaptchaField(
+        widget=widgets.ReCaptchaV2Checkbox(
+            attrs={
+                'data-theme': 'dark',
+                'data-size': 'compact',
+                'language': 'cs'
+            }
+        )
+    )
+    # The ReCaptchaV2Invisible widget,
```
Comment (cobusc) [inline README.rst]: No comma after "widget"

File: README.rst [hunk 8]
```diff
@@ -93,23 +88,57 @@ To allow for runtime specification of keys you can optionally pass the
...
+                'language': 'cs'
+            }
+        )
+    )
+    # The ReCaptchaV2Invisible widget, 
+    # ignores the "data-size" attribute in favor of 'data-size="invisible"'
+
+By default the widgets provided only support a single form with a single widget per page.
+
+However the JavaScript used by the widgets are in a separate template directory, that gets included at the top of the widget templates.
```
Comment (cobusc) [inline README.rst]: "However,"

File: captcha/fields.py [hunk 9]
```diff
@@ -7,77 +7,85 @@
...
+                "captcha.fields.ReCaptchaField.widget"
+                " must be a subclass of captcha.widgets.ReCaptchaBase"
+            )
+
+        # reCAPTCHA fields are always required.
         self.required = True
-        super(ReCaptchaField, self).__init__(*args, **kwargs)
+
+        # Setup instance variables.
+        # TODO: Perhaps this should not default to the test keys, indicate test keys in Docs and error if no keys are present.
```
Comment (cobusc) [inline captcha/fields.py]: 3rd option: log a warning that the test keys are being used.

File: captcha/fields.py [hunk 10]
```diff
@@ -7,77 +7,85 @@
...
+        self.private_key = private_key or getattr(
+            settings, "RECAPTCHA_PRIVATE_KEY", TEST_PRIVATE_KEY)
+        self.public_key = public_key or getattr(
+            settings, "RECAPTCHA_PUBLIC_KEY", TEST_PUBLIC_KEY)
+        self.use_ssl = use_ssl if use_ssl is not None else getattr(
+            settings, "RECAPTCHA_USE_SSL", True)
+
+        # Update widget attrs with data-sitekey.
+        self.widget.attrs.update({"data-sitekey": self.public_key})
```
Comment (cobusc) [inline captcha/fields.py]: Why not just: `self.widget.attrs["data-sitekey"] = self.public_key`?

File: captcha/fields.py [hunk 11]
```diff
@@ -7,77 +7,85 @@
...
 
     def get_remote_ip(self):
         f = sys._getframe()
         while f:
-            if 'request' in f.f_locals:
-                request = f.f_locals['request']
+            if "request" in f.f_locals:
```
Comment (cobusc) [inline captcha/fields.py]: These 2 lines can be condensed to `request = f.f_locals.get("request")`

File: captcha/templates/captcha/includes/js_v2_invisible.html [hunk 12]
```diff
@@ -1,19 +1,21 @@
 {# The provided implementation caters for only one reCAPTCHA on a page. Override this template and its logic as needed. #}
-<script src="https://www.google.com/recaptcha/api.js" async defer></script>
```
Comment (cobusc) [inline captcha/templates/captcha/includes/js_v2_invisible.html]: Weird. Specifying both `async` and `defer` does not make sense, to my knowledge. I'm glad it is gone now.

File: captcha/templates/captcha/includes/js_v2_invisible.html [hunk 13]
```diff
@@ -1,19 +1,21 @@
...
+<script src="https://www.google.com/recaptcha/api.js{% if language %}?hl={{ language }}{% endif %}"></script>
```
Comment (cobusc) [inline captcha/templates/captcha/includes/js_v2_invisible.html]: Should we add `async` to the `script` definition here?

File: captcha/tests/test_fields.py [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+        self.assertIn("var verifyCaptcha_%s = function(e) {" % test_hex, html)
+        self.assertIn('.g-recaptcha[data-widget-uuid="%s"]' % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_v2_invisible_attribute_changes_html(self, mocked_uuid):
+        test_hex = "8b220c54ddb849b8bb59bda5da57baea"
+        mocked_uuid.return_value = test_hex
+        class InvisAttrForm(forms.Form):
```
Comment (cobusc) [inline captcha/tests/test_fields.py]: Whitespace before and after class definition, please.

File: captcha/tests/test_fields.py [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+        self.assertIn('required="True"', html)
+        self.assertIn('data-widget-uuid="%s"' % test_hex, html)
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_default_v2_invisible_html(self, mocked_uuid):
+        test_hex = "72f853eb8b7e4022b808be0f5c3bc297"
+        mocked_uuid.return_value = test_hex
+        class InvisForm(forms.Form):
```
Comment (cobusc) [inline captcha/tests/test_fields.py]: Whitespace before and after class definition, please.

File: captcha/tests/test_fields.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+        self.assertIn('required="True"', html)
+        self.assertIn('data-widget-uuid="%s"' % test_hex, html)
+        self.assertIn('data-sitekey="pubkey"', html)
+        self.assertIn("var onSubmit_%s = function(token) {" % test_hex, html)
+
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_v2_checkbox_attribute_changes_html(self, mocked_uuid):
+        test_hex = "e83ccae286ad4784bd47f7ddc40cfd6f"
+        mocked_uuid.return_value = test_hex
+        class CheckboxAttrForm(forms.Form):
```
Comment (cobusc) [inline captcha/tests/test_fields.py]: Whitespace before and after class definition, please.

File: captcha/tests/test_fields.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+       # Trigger client.submit
+       form.is_valid()
+
+
+class TestWidgets(TestCase):
+    @patch("captcha.widgets.uuid.UUID.hex", new_callable=PropertyMock)
+    def test_default_v2_checkbox_html(self, mocked_uuid):
+        test_hex = "928e8e017b114e1b9d3a3e877cfc5844"
+        mocked_uuid.return_value = test_hex
+        class DefaultCheckForm(forms.Form):
```
Comment (cobusc) [inline captcha/tests/test_fields.py]: Whitespace before and after class definition, please.

File: captcha/tests/test_fields.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+    def test_widget_check(self):
+        with self.assertRaises(ImproperlyConfigured):
+            class ImporperForm(forms.Form):
+                captcha = fields.ReCaptchaField(widget=forms.Textarea)
+
+    @patch("captcha.fields.client.submit")
+    def test_field_instantiate_values(self, mocked_submit):
+        mocked_submit.return_value = RecaptchaResponse(is_valid=True)
+        class NonDefaultForm(forms.Form):
```
Comment (cobusc) [inline captcha/tests/test_fields.py]: Whitespace before and after class definition, please.

File: captcha/widgets.py [hunk 19]
```diff
@@ -1,88 +1,61 @@
...
-
-        try:
-            context = super(ReCaptcha, self).get_context(name, value, attrs)
-        except AttributeError:
-            context = {
-                "widget": {
-                    "attrs": self.build_attrs(attrs)
-                }
-            }
+        # TODO make use of django.utils.translation import get_language
```
Comment (cobusc) [inline captcha/widgets.py]: Are you going to do this?

File: captcha/tests/test_fields.py (hunk 14 above)
Comment (AltusBarry) [reply captcha/tests/test_fields.py]: As per external discussion, JavaScript and HTML improvements can be done at a later date.

File: captcha/templates/captcha/includes/js_v2_invisible.html (hunk 13 above)
Comment (AltusBarry) [reply captcha/templates/captcha/includes/js_v2_invisible.html]: As per external discussion, JavaScript and HTML improvements can be done at a later date.

File: captcha/templates/captcha/includes/js_v2_invisible.html (hunk 12 above)
Comment (AltusBarry) [reply captcha/templates/captcha/includes/js_v2_invisible.html]: As per external discussion, JavaScript and HTML improvements can be done at a later date.

**Comment by cobusc** [comment]
;+1:

**Pull request closed without merging**